django/tasks/utils.py

```diff
@@ -0,0 +1,71 @@
+import inspect
+import json
+import time
+from functools import wraps
+from traceback import format_exception
+
+from django.utils.module_loading import import_string
+
+
+def is_module_level_function(func):
+    if not inspect.isfunction(func) or inspect.isbuiltin(func):
+        return False
+
+    if "<locals>" in func.__qualname__:
+        return False
+
+    return True
+
+
+def json_normalize(obj):
+    """
+    Round-trip encode object as JSON to normalize types.
+    """
+    return json.loads(json.dumps(obj))
+
+
+def retry(*, retries=3, backoff_delay=0.1):
+    """
+    Retry the given code `retries` times, raising the final error.
+
+    `backoff_delay` can be used to add a delay between attempts.
+    """
+
+    def wrapper(f):
+        @wraps(f)
+        def inner_wrapper(*args, **kwargs):
+            for attempt in range(1, retries + 1):
+                try:
+                    return f(*args, **kwargs)
+                except KeyboardInterrupt:
+                    # Let the user ctrl-C out of the program without a retry
+                    raise
+                except BaseException:
+                    if attempt == retries:
+                        raise
+                    time.sleep(backoff_delay)
+
+        return inner_wrapper
+
+    return wrapper
+
+
+def get_module_path(val):
+    return f"{val.__module__}.{val.__qualname__}"
+
+
+def exception_to_dict(exc):
+    return {
+        "exc_type": get_module_path(type(exc)),
+        "exc_args": json_normalize(exc.args),
+        "exc_traceback": "".join(format_exception(type(exc), exc, exc.__traceback__)),
+    }
+
+
+def exception_from_dict(exc_data):
+    exc_class = import_string(exc_data["exc_type"])
+
+    if not inspect.isclass(exc_class) or not issubclass(exc_class, BaseException):
+        raise TypeError(f"{type(exc_class)} is not an exception")
+
+    return exc_class(*exc_data["exc_args"])
```

Comment:
Removing `import_string` isn't really going to happen. It's specifically validated as being an exception, which should avoid most of the even hypothetical issues. 

As a user looking at a failed task, it may be helpful for the application to know why the task failed, in much the same way as a `except` statement.

Reply:
My pitch here would be that a stringified exception would be good enough to know why the task failed, without having serialization/deserialization logic be needed for every kind of exception one might hit (given exceptions will often have rich data). But if I'm the only one with this complaint, I'm not going to stand in the way here